File: Go/src/instrumentation.go
package dheap

// Comparison-count instrumentation for the priority queue (v2.6.0 Phase 2).
//
// Mirrors the contract from TypeScript/src/instrumentation.ts (v2.4.0): count
// *comparisons* (not operations), bucketed by which heap operation triggered
// them. Per the v2.6.0 ROADMAP, the Go mechanism is a nil pointer in the heap:
// pq.stats *Stats is nil by default — overhead in that path is a single
// well-predicted nil check per comparison.
//
// Ownership note: in Go, the *user* owns the Stats storage (typically a
// stack-local var dheap.Stats); the heap stores a *Stats pointer that may be
// nil. Read access is via pq.Stats() or via the original variable; Total() and
// Reset() are nil-safe so pq.Stats().Total() always works regardless of
// whether instrumentation was attached. This is deliberately asymmetric with
// C++, where the heap owns the stats by value via [[no_unique_address]] — a
// design choice driven by each language's idioms and zero-overhead mechanism.
//
// Usage:
//
//	var stats dheap.Stats
//	pq := dheap.New(dheap.Options[Vertex, string]{
//	    D:            4,
//	    Comparator:   dheap.MinBy(func(v Vertex) int { return v.Distance }),
//	    KeyExtractor: func(v Vertex) string { return v.ID },
//	    Stats:        &stats,
//	})
//	// ... use pq ...
//	fmt.Printf("inserts=%d, pops=%d, total=%d\n", stats.Insert, stats.Pop, stats.Total())
//
// Cross-language equivalents:
//   - TypeScript: ComparisonStats from instrumentation.ts (v2.4.0)
//   - C++: TOOLS::ComparisonStats from PriorityQueue.h (v2.6.0)
//   - Rust: StatsCollector trait (Phase 2, planned)
//   - Zig: comptime bool parameter (Phase 2, planned)

// OperationType identifies which public heap method triggered a comparison.
type OperationType int

const (
	OpNone OperationType = iota
	OpInsert
	OpPop
	OpDecreasePriority
	OpIncreasePriority
	OpUpdatePriority
)

// Stats holds per-operation comparison counts. Pass a *Stats via
// Options.Stats to opt in. Counters are public so callers can read them
// directly; Total() and Reset() are convenience methods.
//
// Cross-language equivalents:
//   - C++: TOOLS::ComparisonStats
//   - TypeScript: ComparisonStats
type Stats struct {
	Insert           uint64
	Pop              uint64
	DecreasePriority uint64
	IncreasePriority uint64
	UpdatePriority   uint64

	// currentOp is heap-internal: set by startOperation, cleared by endOperation.
	// Users do not touch this directly.
	currentOp OperationType
}

// Total returns the sum of all five counter buckets.
//
// Nil-safe: returns 0 when called on a nil receiver. This makes
// `pq.Stats().Total()` safe even when no Stats was attached at construction —
// matching the C++ `pq.stats().total()` semantics where the default NoOpStats
// policy always returns 0.
func (s *Stats) Total() uint64 {
	if s == nil {
		return 0
	}
	return s.Insert + s.Pop + s.DecreasePriority + s.IncreasePriority + s.UpdatePriority
}

// Reset zeroes all counters and clears the active-operation tag. The heap
// state is unaffected by this call — the heap and the stats are independent.
//
// Nil-safe: a no-op when called on a nil receiver.
//
// Not safe to call concurrently with heap operations on another goroutine —
// the package as a whole assumes single-goroutine use.
func (s *Stats) Reset() {
	if s == nil {
		return
	}
	*s = Stats{}
}

// startOperation is called by the heap at the entry of each public mutator.
// Unexported: heap-internal hook.
func (s *Stats) startOperation(op OperationType) {
	s.currentOp = op
}

// endOperation is called by the heap at the exit of each public mutator
// (typically via defer).
func (s *Stats) endOperation() {
	s.currentOp = OpNone
}

// countComparison is called by the heap each time the comparator runs.
// Comparisons attributed to OpNone (i.e. outside a bracketed operation) are
// dropped — there shouldn't be any in normal flow.
func (s *Stats) countComparison() {
	switch s.currentOp {
	case OpInsert:
		s.Insert++
	case OpPop:
		s.Pop++
	case OpDecreasePriority:
		s.DecreasePriority++
	case OpIncreasePriority:
		s.IncreasePriority++
	case OpUpdatePriority:
		s.UpdatePriority++
	}
}

File: Rust/src/instrumentation.rs
//! Comparison-count instrumentation for the priority queue (v2.6.0 Phase 2).
//!
//! Mirrors the contract from `TypeScript/src/instrumentation.ts` (v2.4.0):
//! count *comparisons* (not operations), bucketed by which heap operation
//! triggered them. Per the v2.6.0 ROADMAP, the Rust mechanism is **generic over
//! a `StatsCollector` trait**, with monomorphization providing zero overhead
//! when the default `NoOpStats` is used.
//!
//! # Usage
//!
//! ```rust
//! use d_ary_heap::{MinBy, PriorityQueue, StatsCollector};
//!
//! let mut pq = PriorityQueue::with_stats(2, MinBy(|x: &i32| *x)).unwrap();
//! pq.insert(5);
//! pq.insert(3);
//! let _ = pq.pop();
//! let stats = pq.stats();
//! println!("inserts: {}, pops: {}, total: {}",
//!     stats.insert(), stats.pop(), stats.total());
//! ```
//!
//! # Ownership note
//!
//! In Rust, the heap **owns** the stats by value. With the default
//! `NoOpStats` (a zero-sized type), the `stats` field collapses to zero bytes
//! via Rust's standard ZST layout — no attribute is needed (unlike C++'s
//! `[[no_unique_address]]`). With `ComparisonStats` attached (via the
//! `InstrumentedPriorityQueue` alias), the heap holds a small fixed-size
//! struct of `Cell<u64>` counters. This is deliberately asymmetric with Go,
//! where the user owns the storage and the heap holds a `*Stats` pointer.
//!
//! # Cross-language equivalents
//!
//! - C++: `TOOLS::ComparisonStats` and `TOOLS::OperationType` in `PriorityQueue.h`
//! - Go: `dheap.Stats` and `dheap.OperationType` in `instrumentation.go`
//! - TypeScript: `ComparisonStats` from `instrumentation.ts` (v2.4.0)
//! - Zig: comptime bool parameter (Phase 2, planned)

use std::cell::Cell;

/// Identifies which public heap method triggered a comparison.
///
/// `#[non_exhaustive]` because future operations (e.g., `Heapify`) can be added
/// without breaking exhaustive matchers in downstream code; the only exhaustive
/// matcher in this crate lives inside `count_comparison()` and stays correct
/// because it owns the enum.
#[non_exhaustive]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum OperationType {
    /// No operation in flight; comparisons attributed to `None` are dropped.
    #[default]
    None,
    /// `insert` / `insert_many`.
    Insert,
    /// `pop` (and `pop_many` via per-call delegation).
    Pop,
    /// `decrease_priority` and `decrease_priority_by_index`.
    DecreasePriority,
    /// `increase_priority` and `increase_priority_by_index`.
    IncreasePriority,
    /// `update_priority` and `update_priority_by_index`.
    UpdatePriority,
}

/// Trait the heap drives during operations.
///
/// All methods take `&self` (not `&mut self`) so the heap's `&self` query
/// methods (`best_child_position`, etc.) can update the active operation /
/// counters via interior mutability. `ComparisonStats` uses `Cell<T>` for that;
/// `NoOpStats` has nothing to mutate.
pub trait StatsCollector {
    /// Mark an operation as in-flight; subsequent `count_comparison` calls
    /// attribute to the matching bucket.
    fn start_operation(&self, op: OperationType);

    /// Mark the in-flight operation as complete; subsequent comparisons (if
    /// any) attribute to `OperationType::None` and are dropped.
    fn end_operation(&self);

    /// Increment the bucket of the currently-active operation.
    fn count_comparison(&self);

    /// Sum of all five operation buckets. With `NoOpStats` this returns 0.
    fn total(&self) -> u64;

    /// Zero all counters and reset the active-operation tag.
    fn reset(&self);
}

/// Zero-sized policy: every method is an empty no-op.
///
/// With the default `S = NoOpStats`, monomorphization specializes each call
/// site to a no-op, the `stats` field of the heap is zero bytes (Rust ZST
/// layout), and the bracketing `OpGuard`'s `Drop` body is empty — i.e. zero
/// runtime cost.
#[derive(Default, Debug, Clone, Copy)]
pub struct NoOpStats;

impl StatsCollector for NoOpStats {
    #[inline]
    fn start_operation(&self, _op: OperationType) {}
    #[inline]
    fn end_operation(&self) {}
    #[inline]
    fn count_comparison(&self) {}
    #[inline]
    fn total(&self) -> u64 {
        0
    }
    #[inline]
    fn reset(&self) {}
}

/// Real comparison-count policy. Five buckets indexed by the active
/// `OperationType`, plus the active-operation tag.
///
/// All counters are `Cell<u64>` (interior mutability), allowing the trait
/// methods to take `&self`. Hot-path cost when attached: a single `Cell::get`
/// + branch on the active op, then `Cell::set`.
#[derive(Default, Debug, Clone)]
pub struct ComparisonStats {
    insert: Cell<u64>,
    pop: Cell<u64>,
    decrease_priority: Cell<u64>,
    increase_priority: Cell<u64>,
    update_priority: Cell<u64>,
    current_op: Cell<OperationType>,
}

impl ComparisonStats {
    /// Comparisons triggered by `insert` / `insert_many`.
    #[must_use]
    pub fn insert(&self) -> u64 {
        self.insert.get()
    }
    /// Comparisons triggered by `pop` (and `pop_many` via per-call delegation).
    #[must_use]
    pub fn pop(&self) -> u64 {
        self.pop.get()
    }
    /// Comparisons triggered by `decrease_priority` / `decrease_priority_by_index`.
    #[must_use]
    pub fn decrease_priority(&self) -> u64 {
        self.decrease_priority.get()
    }
    /// Comparisons triggered by `increase_priority` / `increase_priority_by_index`.
    #[must_use]
    pub fn increase_priority(&self) -> u64 {
        self.increase_priority.get()
    }
    /// Comparisons triggered by `update_priority` / `update_priority_by_index`.
    #[must_use]
    pub fn update_priority(&self) -> u64 {
        self.update_priority.get()
    }
}

impl StatsCollector for ComparisonStats {
    fn start_operation(&self, op: OperationType) {
        self.current_op.set(op);
    }

    fn end_operation(&self) {
        self.current_op.set(OperationType::None);
    }

    fn count_comparison(&self) {
        match self.current_op.get() {
            OperationType::Insert => self.insert.set(self.insert.get() + 1),
            OperationType::Pop => self.pop.set(self.pop.get() + 1),
            OperationType::DecreasePriority => {
                self.decrease_priority.set(self.decrease_priority.get() + 1);
            }
            OperationType::IncreasePriority => {
                self.increase_priority.set(self.increase_priority.get() + 1);
            }
            OperationType::UpdatePriority => {
                self.update_priority.set(self.update_priority.get() + 1);
            }
            // INDEX: None means no operation is in flight; comparisons here are
            // unattributed and dropped (matches C++/Go semantics).
            OperationType::None => {}
        }
    }

    fn total(&self) -> u64 {
        self.insert.get()
            + self.pop.get()
            + self.decrease_priority.get()
            + self.increase_priority.get()
            + self.update_priority.get()
    }

    fn reset(&self) {
        self.insert.set(0);
        self.pop.set(0);
        self.decrease_priority.set(0);
        self.increase_priority.set(0);
        self.update_priority.set(0);
        self.current_op.set(OperationType::None);
    }
}
